src/storage/src/hummock/sstable/xor_filter.rs

```diff
@@ -136,9 +142,126 @@ impl FilterBuilder for Xor8FilterBuilder {
     }
 }
 
+pub struct BlockedXor16FilterBuilder {
+    current: Xor16FilterBuilder,
+    data: Vec<u8>,
+    block_count: usize,
+}
+
+const BLOCK_FILTER_CAPACITY: usize = 16 * 1024; // 16KB means 2K key count.
+
+impl BlockedXor16FilterBuilder {
+    pub fn is_kv_count_too_large(kv_count: usize) -> bool {
+        kv_count > MAX_KV_COUNT_FOR_XOR16
+    }
+
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            current: Xor16FilterBuilder::new(BLOCK_FILTER_CAPACITY),
+            data: Vec::with_capacity(capacity),
+            block_count: 0,
+        }
+    }
+}
+
+impl FilterBuilder for BlockedXor16FilterBuilder {
+    fn add_key(&mut self, key: &[u8], table_id: u32) {
+        self.current.add_key(key, table_id)
+    }
+
+    fn finish(&mut self, _memory_limiter: Option<Arc<MemoryLimiter>>) -> Vec<u8> {
+        // Add footer to tell which kind of filter. 254 indicates a xor8 filter.
+        self.data.put_u32_le(self.block_count as u32);
+        self.data.put_u8(FOOTER_BLOCKED_XOR16);
+        std::mem::take(&mut self.data)
+    }
+
+    fn approximate_len(&self) -> usize {
+        self.current.approximate_len() + self.data.len()
+    }
+
+    fn create(_fpr: f64, capacity: usize) -> Self {
+        BlockedXor16FilterBuilder::new(capacity)
+    }
+
+    fn switch_builder(&mut self) {
+        let block = self.current.finish(None);
+        self.data.put_u32_le(block.len() as u32);
+        self.data.extend(block);
+        self.block_count += 1;
+    }
+
+    fn approximate_building_memory(&self) -> usize {
+        0
```

Comment:
How about Keeping the memory request when finish BlockedXor16FilterBuilder like other XorFilter.

Even if the number of keys in a single block is small, but the amount of concurrency used by caller is uncertain, we can still make a guarantee.

Reply:
But this method would never be used..